### PR TITLE
Update to Boogie 2.12.1

### PR DESCRIPTION
**Pull request by keyboardDrummer**
Update to Boogie 2.12.1

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/DafnyLanguageServer/Language/DafnyProgramVerifier.cs [hunk 1]
```diff
@@ -25,10 +25,14 @@ public class DafnyProgramVerifier : IProgramVerifier {
...
     private readonly VerifierOptions options;
     private readonly SemaphoreSlim mutex = new(1);
+    private static readonly VerificationResultCache cache = new();
+
+    DafnyOptions Options => DafnyOptions.O;
 
     private DafnyProgramVerifier(ILogger<DafnyProgramVerifier> logger, VerifierOptions options) {
       this.logger = logger;
       this.options = options;
+
```
Comment (MikaelMayer) [inline Source/DafnyLanguageServer/Language/DafnyProgramVerifier.cs:35]:
```suggestion
```

File: Source/Dafny/DafnyOptions.cs [hunk 2]
```diff
@@ -17,6 +17,12 @@ namespace Microsoft.Dafny {
   public class DafnyOptions : Bpl.CommandLineOptions {
     private ErrorReporter errorReporter;
 
+    public new static DafnyOptions FromArguments(params string[] arguments) {
```
Comment (MikaelMayer) [inline Source/Dafny/DafnyOptions.cs:20]: While looking at the tests, the call `DafnyOptions.FromArguments()` felt odd to me, because I was looking for "where are the static/global arguments defined that this function pulls from?". I did not realize immediately that it's the empty parameter list that was the arguments, until I came back to this function.

I don't have a strong opinion on that (consider it optional), but I would have felt much more comfortable with a name such as `Create` so that  `DafnyOption.Create()` really feels like it's not depending on any external context. It would still feel natural that: `DafnyOptions.Create("--compile:0"). is actually behaving like command-line arguments.

**Review by MikaelMayer — APPROVED** [review]
Looks good to me, two suggestions to consider to improve this PR.
I'm looking forward to have this version of Boogie that fixes many concurrency problems.

File: Source/DafnyLanguageServer/Workspace/TextDocumentLoader.cs [hunk 3]
```diff
@@ -51,6 +52,7 @@ public class TextDocumentLoader : ITextDocumentLoader {
       this.ghostStateDiagnosticCollector = ghostStateDiagnosticCollector;
       this.notificationPublisher = notificationPublisher;
       this.loggerFactory = loggerFactory;
+      this.logger = loggerFactory.CreateLogger(typeof(TextDocumentLoader));
```
Comment (MikaelMayer) [inline Source/DafnyLanguageServer/Workspace/TextDocumentLoader.cs:55]:
```suggestion
      this.logger = loggerFactory.CreateLogger<TextDocumentLoader>();
```

File: Source/DafnyLanguageServer/Workspace/TextDocumentLoader.cs [hunk 4]
```diff
@@ -34,6 +34,7 @@ public class TextDocumentLoader : ITextDocumentLoader {
     private readonly ILoggerFactory loggerFactory;
     private readonly BlockingCollection<Request> requestQueue = new();
     private readonly IOptions<DafnyPluginsOptions> dafnyPluginsOptions;
+    private readonly ILogger logger;
```
Comment (MikaelMayer) [inline Source/DafnyLanguageServer/Workspace/TextDocumentLoader.cs:37]:
```suggestion
    private readonly ILogger<TextDocumentLoader> logger;
```

File: Source/DafnyLanguageServer/Workspace/TextDocumentLoader.cs (hunk 3 above)
Comment (MikaelMayer) [reply Source/DafnyLanguageServer/Workspace/TextDocumentLoader.cs:55]: to match what is done in that same file with the loggerFactory.

File: Source/DafnyPipeline.Test/IntraMethodVerificationStability.cs [hunk 5]
```diff
@@ -162,63 +165,65 @@ class C {
...
       var temp1 = directory + "/proverLog";
       testOutputHelper.WriteLine("proverLog: " + temp1);
       options.ProverLogFilePath = temp1;
-      var engine = ExecutionEngine.CreateWithoutSharedCache(options);
-      foreach (var boogieProgram in boogiePrograms) {
-        Main.BoogieOnce(engine, "", "", boogieProgram, "programId", out _, out var outcome);
-        testOutputHelper.WriteLine("outcome: " + outcome);
-        foreach (var proverFile in Directory.GetFiles(directory)) {
-          yield return File.ReadAllText(proverFile);
+      using var engine = ExecutionEngine.CreateWithoutSharedCache(options);
```
Comment (MikaelMayer) [inline Source/DafnyPipeline.Test/IntraMethodVerificationStability.cs:217]: +1 on using 'using' to dispose the execution engine!

File: Source/DafnyTestGeneration/ProgramModification.cs [hunk 6]
```diff
@@ -85,7 +85,7 @@ public class ProgramModification {
       engine.Inline(program);
       var log = Utils.CaptureConsoleOutput(
         () => engine.InferAndVerify(program,
-          new PipelineStatistics(), uniqueId,
+          new PipelineStatistics(), null,
```
Comment (MikaelMayer) [inline Source/DafnyTestGeneration/ProgramModification.cs:88]: Just to confirm: is this change necessary?

File: Source/DafnyTestGeneration/ProgramModification.cs [hunk 7]
```diff
@@ -75,15 +75,16 @@ public class ProgramModification {
       var oldOptions = DafnyOptions.O;
       var options = SetupOptions(procedure);
       DafnyOptions.Install(options);
```
Comment (MikaelMayer) [inline Source/DafnyTestGeneration/ProgramModification.cs:77]: Why do we still need to install options when you are passing options everywhere else?

**Review by MikaelMayer — APPROVED** [review]
Great job. Just a non-blocking question.

File: Source/DafnyTestGeneration/ProgramModification.cs (hunk 7 above)
Comment (keyboardDrummer) [reply Source/DafnyTestGeneration/ProgramModification.cs:77]: In Dafny we still rely on `DafnyOptions.O` a lot which needs to be installed. In a future PR we'll remove `DafnyOptions.O` and references to it.
